orttraining/orttraining/python/training/ortmodule/_training_manager.py

```diff
@@ -37,8 +37,7 @@ def execution_session_run_forward(execution_session, onnx_model, device, *inputs
         # have the need for passing IOBinding.
         state = C.PartialGraphExecutionState()
         forward_inputs = C.OrtValueVector()
-        for input in inputs:
-            forward_inputs.append(_utils._ortvalue_from_torch_tensor(input))
+        forward_inputs.extend([_utils._ortvalue_from_torch_tensor(input) for input in inputs])
```

Comment:
`C.OrtValueVector()` is a wrapped `std::vector<OrtValue>`... is there a way we could call `reserve(len(inputs))` on the wrapped vector?

(If we could, I think `append()` (i.e. `push_back()`) would be more efficient than building a list to pass it to `extend()` as  here.)

Reply:
Seems like pybind had support for C++ style functions https://github.com/pybind/pybind11/blob/e08a58111dbea38d667b209f7543864d51a3b185/include/pybind11/stl_bind.h#L506 which has been deprecated.
I am not sure if there is a way to extend ```bind_vector``` implementation in some way.

Reply:
Boo :(.

In principle, we could add our own specialization of the opaque wrapper with just the methods we need. (I don't think the vector binding uses any private/internal pybind APIs.)

I'd propose the following for this part of the change:
1. If it makes things go faster, let's just merge it!
2. Let's implement a replacement that optimizes for exactly the case we have here, namely:
    * Pre-reserve the exact length.
    * Add an `extend()` method that performs the conversion from torch `Tensor` to `ORTValue` in C++, without iterating over `inputs` in Python at all.

ISTR that the `_ortvalue_from_torch_tensor()` call can be costly here. Another option we have would be to preconvert all the `inputs` that are the same from one step to the next (i.e. the parameters, buffers).

Reply:
Reading more, it appears that the pybind ```extend``` on bounded vectors does something similar to what we want to achieve:
https://github.com/pybind/pybind11/blob/e08a58111dbea38d667b209f7543864d51a3b185/include/pybind11/stl_bind.h#L154

So, I am assuming that doing this explicitly in python would then not be necessary. Please correct me if I am mistaken.

Reply:
I have added a few methods (such as reserve) that will help us.
As far as performing the conversion from torch ```Tensor``` to ```ORTValue``` in C++ goes, I think since we do not have the torch header files in C++ (and presumably do not want to take dependency on that), I am not sure if this can be achieved on the C++ side.

Reply:
@thiagocrepaldi just mentioned to me that this can be achieved using the dlpack. Let me give that a try.

Reply:
I am realizing that the iteration over the tensor must happen on the python side since the ```torch.tensor``` must be converted to ```dlpack```. The ```C++``` side is only able to understand either a ```dlpack``` or an ```OrtValue```.

I am adding the ability to retrieve a dlpack directly from the vector though.